### DISCUSSION TITLE
Undefined weak symbols

### DISCUSSION THREAD

**Discussion by ilg-ul** [iains/gcc-darwin-arm64#133] [Q&A] [answered]
On Linux, if a weak symbol is not defined but used in an `if` statement, it returns zero, and so the program knows at runtime that the symbol is not defined.

On macOS, according to my tests, the Apple linker does not accept undefined weak symbols.

Can you confirm that this is the normal behaviour of the current Apple linker?

Is there any linker option that could change this behaviour? (I could not find anything related)

Thank you!

**Comment by iains** [comment, marked as answer]
Default ELF linker semantics are different from default Mach-O linker semantics.  ELF allows any undefined symbols, Mach-O does not (note that there are ELF linker options to make it reject missing symbols and, as I am going to write below, options to allow missing symbols on Mach-O).

In both cases, weak referenced symbols do operate _**at runtime**_ in the same way (if the symbol is not present in the dynamically-linked exe, it will be 0).

For Mach-O, using current linkers, you have two ways to deal with this;

1. you tell the linker that it's OK for a set of symbols to be undefined at static-link time `-U,_symbol` .. or (if there are many) by providing the list in a file [ man ld ].
2. you tell the linker that it's OK for _any_ symbol to be undefined at static link time `-undefined dynamic_lookup`

the second is a large sledgehammer and will not only allow the symbols you _want_ to be undefined, but any others that get accidentally misspelled etc. 

There are times when the second is useful (mostly in implementing plugins or when too lazy to specify the correct set).

additional: dynamic lookup is incompatible with some relocation mechanisms used on Arm64 (according to linker warnings) .. so perhaps it is wise to consider mechanisms to avoid that sledgehammer...

**Reply by ilg-ul** [reply]
Thank you for the detailed answer.

Personally I would not use undefined weaks at all, but most of my tests were inherited from Linux, and I'm trying to make use of them on all platforms.

To clarify my actual dilema: does this test (if undefined weak symbols return zero) make any sense on macOS? Initially I left it enabled, but marked it as expected to fail. Now I completely disabled it when running on macOS, since it is known to always fail.

Any advice?

**Reply by iains** [reply]
> To clarify my actual dilema: does this test (if undefined weak symbols return zero) make any sense on macOS? Initially I left it enabled, but marked it as expected to fail. Now I completely disabled it when running on macOS, since it is known to always fail.

Providing you mean 'weak reference' (see below)....

Yes, they make perfect sense on macOS - they are used (for example) when an application is linked with an optional library - when the library symbols are present, they can be used, when absent either the functionality is missing - or some fallback is used.   The other use is when an executable might be used on different OS versions some of which have an implementation of a function, and some don't.

NOTE: this is a **_different_** use-case than making a symbol a weak definition to satisfy ODR constraints (in that case the symbol is not optional - it would not be marked as a weak reference).

**Reply by iains** [reply]
> NOTE: this is a **_different_** use-case than making a symbol a weak definition to satisfy ODR constraints (in that case the symbol is not optional - it would not be marked as a weak reference).

I suppose, for completeness I should also mention a third use-case for weak.  That is when a weak _**definition**_ is expected to be overridden at static link-time by a non-weak one.  This is used on macOS to implement the user-replaceable allocators in c++ for example.  Once again there is no weak **_reference_**  here; the symbol must be provided - but if a non-weak one _is_ provided, then that will be used instead of the weak definition (in the replaceable allocators case, the latter is in the c++ library).

**Reply by ilg-ul** [reply]
In the embedded world where I come from, weak is used in some frameworks to define default functions which can be reimplemented by the application. This creates a lot of issues since such code cannot be modularised easily, and the whole build must be monolithic.

The use cases that you describe are somehow similar, but probably slightly more complex, and I would say not covered by the tests I use.

The test I'm referring to is made of 3 files:

main-weak.c:
```
#include <stdio.h>

__attribute__((weak)) void func(void);
void dummy(void);

int value = 1;
extern int expected;

int main(int argc, char **argv) {
    if (func)
        func();
    dummy();
    if (value != expected) {
        printf("expected %d, got %d\n", expected, value);
        return 1;
    }
    return 0;
}
```

expected1.c:
```
int expected = 1;
```

dummy.c:
```
void dummy(void) {
}
```

It is part of a suite of tests, here the interesting part is the missing `func`, which on Linux returns 0, while on macOS fails to link.

**Reply by iains** [reply]
> In the embedded world where I come from, weak is used in some frameworks to define default functions which can be reimplemented by the application. This creates a lot of issues since such code cannot be modularised easily, and the whole build must be monolithic.

Design of frameworks and/or applications is outside the scope of the "implementation" if the toolchain &c.     The toolchain provides facilities - it's up to the systems' engineers to decide how to use those.

> It is part of a suite of tests, here the interesting part is the missing `func`, which on Linux returns 0, while on macOS fails to link.

Which is the question that is answered above; this is only a consequence of differing defaults for the static linker between (most) ELF and Mach-O based systems, it makes no difference to the actual runtime behaviour.

**Reply by iains** [reply]
> In the embedded world where I come from, weak is used in some frameworks to define default functions which can be reimplemented by the application. This creates a lot of issues since such code cannot be modularised easily, and the whole build must be monolithic.

Although - as you describe it here - this is the "replaceable function" case (3) not the "maybe provided, maybe not" (1) - although the same functionality could probably be designed to use the "maybe provided" mechanism - in which case the build would not need to be monolithic.

... but now we're digressing :)
... the answer remains "macOS can do the same as Linux, you just need to provide the relevant information to the static linker".  There are tests in the GCC testsuite that do exactly this.

**Reply by ilg-ul** [reply]
The point here is neither the design of embedded frameworks (which are a mess) nor the need for monolithic builds (which is the usual approach for embedded).

The question is related to this specific test, does it make any sense to enable it for GCC on macOS? Is it of any use? (for me it looks like a nuisance, since I had to explicitly mark it as expected to fail).

**Reply by iains** [reply]
> The question is related to this specific test, does it make any sense to enable it for GCC on macOS? Is it of any use? (for me it looks like a nuisance, since I had to explicitly mark it as expected to fail).

I'm not sure why it is not appropriate to fix the test to work as intended? - all you need to do is to add -Wl,-U,_symbolname to the link flags - for a single missing symbol.

**Reply by ilg-ul** [reply]
> all you need to do is to add -Wl,-U,_symbolname

That was useful, I added `-Wl,-U,_func` and the test passed.

Thank you, I think the question was answered.
